clients/venice-producer/src/main/java/com/linkedin/venice/producer/online/OnlineVeniceProducer.java

```diff
@@ -0,0 +1,120 @@
+package com.linkedin.venice.producer.online;
+
+import static com.linkedin.venice.ConfigKeys.CLIENT_PRODUCER_SCHEMA_REFRESH_INTERVAL_SECONDS;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.linkedin.venice.client.schema.RouterBackedSchemaReader;
+import com.linkedin.venice.client.store.ClientConfig;
+import com.linkedin.venice.client.store.InternalAvroStoreClient;
+import com.linkedin.venice.controllerapi.VersionCreationResponse;
+import com.linkedin.venice.exceptions.VeniceException;
+import com.linkedin.venice.producer.AbstractVeniceProducer;
+import com.linkedin.venice.producer.VeniceProducer;
+import com.linkedin.venice.schema.SchemaReader;
+import com.linkedin.venice.service.ICProvider;
+import com.linkedin.venice.utils.ObjectMapperFactory;
+import com.linkedin.venice.utils.RetryUtils;
+import com.linkedin.venice.utils.Utils;
+import com.linkedin.venice.utils.VeniceProperties;
+import io.tehuti.metrics.MetricsRepository;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.Collections;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+
+/**
+ * An implementation of a VeniceProducer suitable for online applications. This producer only supports producing data to
+ * a store's RT topic.
+ *
+ * @see AbstractVeniceProducer
+ * @see VeniceProducer
+ */
+public class OnlineVeniceProducer<K, V> extends AbstractVeniceProducer<K, V> {
+  private static final ObjectMapper OBJECT_MAPPER = ObjectMapperFactory.getInstance();
+
+  private final String storeName;
+  private final InternalAvroStoreClient<K, V> storeClient;
+
+  private final SchemaReader schemaReader;
+  private final ICProvider icProvider;
+
+  OnlineVeniceProducer(
+      InternalAvroStoreClient<K, V> storeClient,
+      SchemaReader kmeSchemaReader,
+      VeniceProperties producerConfigs,
+      MetricsRepository metricsRepository,
+      ICProvider icProvider) {
+    this.storeClient = storeClient;
+    this.storeName = storeClient.getStoreName();
+    this.icProvider = icProvider;
+
+    Duration schemaRefreshPeriod;
```

Comment:
Do we want the schema reader on the producer to auto refresh? Does this mean registering a bad user schema could potentially break things if online writers pick it up automatically?

Reply:
In online producers, `SchemaReader` is used to fetch the schema ids for data that a user has provided. We want to reduce the number of times that a request to write data has to wait since these are online applications.

Since the SchemaReader object is owned by the producer, this will not leak to other parts of the code that use `SchemaReader`

Reply:
Right, perhaps I should rephrase the question. In the online producers world, do we want users to pick the writer schema similar to the offline and stream world today or always use the latest schema. The later will come with a risk which is bad user schema could break online write instantly and doesn't really give user a chance to canary writes using the new schema.

Reply:
We want users to provide whichever writer schema they want like Samza jobs. For partial updates, we need to pick the latest schema (and the primary reason why we need the schema refresh). This will be a superset schema since superset schemas are always enabled for WC stores

Reply:
Got it, thanks for the explanation.